### PR TITLE
Log Levels

### PR DESCRIPTION
**Pull request by misteroneill**
## Description
This adds support for log levels to video.js' logger. Log levels act as a whitelist of logger method(s) that are allowed. The following levels are defined here:

* `off`: Matches none. Any value that can be cast to `false` will have this effect. The most restrictive.
* `all` (default): Matches all Video.js-provided functions (`log`, `log.warn`, and `log.error`).
* `warn`: Matches `log.warn` and `log.error` calls.
* `error`: Matches only `log.error` calls.

Also, this improves the `log.history` handling with the ability to `clear()` it as well as toggle it off and on using `disable()` and `enable()`.

## Specific Changes proposed
- Adds the `log.level()` method which is a getter/setter for log levels.
- Adds the `log.levels` object as a container for the various logging levels.
- **BREAKING CHANGE:** `log.history` is now a function that returns a shallow clone of the internal history array rather than the array itself.
- Adds the `log.history.clear`, `log.history.disable`, `log.history.enable` methods.

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [ ] If necessary, more likely in a feature request than a bug fix
  - [x] Change has been verified in an actual browser (Chome, Firefox, IE)
  - [x] Unit Tests updated or fixed
  - [x] Docs/guides updated
- [ ] Reviewed by Two Core Contributors

### CODE REVIEW

File: src/js/utils/log.js [hunk 1]
```diff
@@ -22,17 +25,19 @@ let log;
...
 
-    // add the type to the front of the message when it's not "log"
+    // Add the type to the front of the message when it's not "log".
     args.unshift(type.toUpperCase() + ':');
   }
 
-  // add to history
-  log.history.push(args);
+  // Add a clone of the args at this point to history.
+  log.history.push([].concat(args));
```
Comment (misteroneill) [inline src/js/utils/log.js]: Changed this to a clone because I think the intent was for `VIDEOJS:` to _not_ get added to history, but it didn't work that way.

File: src/js/utils/log.js [hunk 2]
```diff
@@ -8,6 +8,9 @@ import {isObject} from './obj';
 
 let log;
 
+// This is the private tracking variable for logging level.
+let level = 'debug';
```
Comment (misteroneill) [inline src/js/utils/log.js]: I guess I'm not 100% sure on this one either. I wonder if it makes more sense to turn logging off and make it opt-in for 6.0?

Comment (gkatsev) [inline src/js/utils/log.js]: Maybe add an `info` and default it to that? `info` would be just `videojs.log`.

File: src/js/utils/log.js [hunk 3]
```diff
@@ -93,6 +99,76 @@ log = function(...args) {
...
+ * - `debug` (default): Matches only Video.js-provided functions (`log`,
+ *   `log.warn`, and `log.error`).
+ * - `warn`: Matches `log.warn` and `log.error` calls.
+ * - `error`: Matches only `log.error` calls.
+ *
+ * @type {Object}
+ */
+log.levels = {};
+
+Object.defineProperties(log.levels, {
```
Comment (gkatsev) [inline src/js/utils/log.js]: this will break in IE8.

Comment (misteroneill) [inline src/js/utils/log.js]: D'oh. I guess they can be normal properties.

File: src/js/utils/log.js (hunk 2 above)
Comment (misteroneill) [inline src/js/utils/log.js]: So it would show `log` calls but not `warn` or `error`?

Comment (gkatsev) [inline src/js/utils/log.js]: Yup.

File: src/js/utils/log.js (hunk 3 above)
Comment (misteroneill) [inline src/js/utils/log.js]: Fair point. Well, we can also simply have it use normal properties for browsers that don't support `Object.defineProperties`.

File: src/js/utils/log.js (hunk 2 above)
Comment (misteroneill) [inline src/js/utils/log.js]: That's not something I would've ever considered to be a use-case, honestly.

**Comment by misteroneill** [comment]
I updated JSDoc, but it's not perfectly correct because this structure (function with functions as properties) doesn't really fit within their OO worldview.

File: src/js/utils/events.js [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
  * @module events
- *
- * Event System (John Resig - Secrets of a JS Ninja http://jsninja.com/)
- * (Original book version wasn't completely usable, so fixed some things and made Closure Compiler compatible)
- * This should work very similarly to jQuery's events, however it's based off the book version which isn't as
- * robust as jquery's, so there's probably some differences.
  */
 
+/*
+Event System (John Resig - Secrets of a JS Ninja http://jsninja.com/)
```
Comment (BrandonOCasey) [inline src/js/utils/events.js]: this should probably be reverted, since it is fixed in #3854

File: src/js/utils/log.js [hunk 5]
```diff
@@ -76,32 +83,124 @@ export const logByType = (type, args, stringify = !!IE_VERSION && IE_VERSION < 1
...
+} else {
+  log.levels.DEFAULT = level;
+}
+
+/**
+ * Get or set the current logging level. If a string matching a key from
+ * {@link log.levels} is provided, acts as a setter. Regardless of argument,
+ * returns the current logging level.
+ *
+ * @param  {String} [lvl]
```
Comment (BrandonOCasey) [inline src/js/utils/log.js]: should be lowercase `{string}` rather than uppercase

File: src/js/utils/log.js [hunk 6]
```diff
@@ -8,9 +8,13 @@ import {isObject} from './obj';
 
 let log;
 
+// This is the private tracking variable for logging level.
+let level = 'all';
```
Comment (BrandonOCasey) [inline src/js/utils/log.js]: It may be better to set the default log level in the same location  as the check for `window.VIDEOJS_DEFAULT_LOG_LEVEL`

File: src/js/utils/log.js [hunk 7]
```diff
@@ -76,32 +83,124 @@ export const logByType = (type, args, stringify = !!IE_VERSION && IE_VERSION < 1
...
+ *
+ * @type {Object}
+ */
+log.levels = {};
+
+if (Object.defineProperties) {
+  Object.defineProperties(log.levels, {
+    all: {
+      enumerable: true,
+      value: 'log|warn|error'
```
Comment (BrandonOCasey) [inline src/js/utils/log.js]: typical log level implementations that I have seen use number rather rather than regex matches to see if they should log. I think that this works fine here and it does allow us to only log from one level if we want to, but I figured that I would throw in another idea.

Here are the log levels I see:
* 0 - Off
* 1 - Error
* 2 - Warn
* 3 - log

Typically you set your log level to the number or string representation of a level, so 3 or `log` by default. Then everything below that level is logged. So if the log level is 3 we log everything, if the log level is 2 we log only warnings and errors, if the log level is 1 we only log errors, and finally if the log level is 0 we log nothing at all.

Comment (misteroneill) [inline src/js/utils/log.js]: That was my original implementation, but the problem with numerically-indexed levels like that is that they are really inflexible and difficult to expand. For example, @gkatsev had the suggestion of a log level that was `log` _only_ (not `warn` or `error`). That doesn't fit neatly in this sort of sequential system.

Comment (BrandonOCasey) [inline src/js/utils/log.js]: ahh ok, just wanted to make sure the more traditional system was considered, but I do see the benefits here.

File: docs/guides/faq.md [hunk 8]
```diff
@@ -74,6 +74,13 @@ help but are outside of the scope of video.js.
 
 For content that must be highly secure [videojs-contrib-eme][eme] adds DRM support.
 
+## Q: Can I turn off video.js logging?
+Yes! This can be achieved by adding the following code _before_ including video.js:
+
+```js
+window.VIDEOJS_DEFAULT_LOG_LEVEL = 'off';
```
Comment (gkatsev) [inline docs/guides/faq.md]: can we make this be settable per-player also, not just globally?

**Comment by misteroneill** [comment]
Yeah, it could be a minor. I'll make the other change.

File: docs/guides/faq.md (hunk 8 above)
Comment (misteroneill) [inline docs/guides/faq.md]: I think the only way to do this would be to really refactor the logging to create new console wrappers for each player (alongside the one for video.js itself). Probably not worth doing for a minor.

**Comment by misteroneill** [comment]
As discussed offline, we are going to refactor this somewhat to allow for per-player logging and log-levels, which means it will become a major update and roll into 6.0.

File: docs/faq.md [hunk 9]
```diff
@@ -153,6 +160,15 @@ the [plugins guide][plugin-guide] for more information.
 
 See the [video.js github wiki][skins-list].
 
+## Q: Can I disable the `<style>` elements video.js adds by default?
```
Comment (BrandonOCasey) [inline docs/faq.md]: Should this be in another PR?

Comment (BrandonOCasey) [inline docs/faq.md]: Should this be in another PR?

File: src/js/utils/log.js [hunk 10]
```diff
@@ -76,32 +83,127 @@ export const logByType = (type, args, stringify = !!IE_VERSION && IE_VERSION < 1
...
+      enumerable: true,
+      value: ''
+    },
+    warn: {
+      enumerable: true,
+      value: 'warn|error'
+    }
+  });
+} else {
+  log.levels = {
```
Comment (BrandonOCasey) [inline src/js/utils/log.js]: what about making this the default and overriding it if `Object.defineProperties` exists? We could also loop through the original log levels to make this code smaller.

File: src/js/utils/log.js [hunk 11]
```diff
@@ -76,32 +83,127 @@ export const logByType = (type, args, stringify = !!IE_VERSION && IE_VERSION < 1
...
+    off: '',
+    warn: 'warn|error'
+  };
+}
+
+// Determine the default logging level.
+level = 'all';
+
+if (window.VIDEOJS_DEFAULT_LOG_LEVEL &&
+    log.levels.hasOwnProperty(window.VIDEOJS_DEFAULT_LOG_LEVEL)) {
```
Comment (BrandonOCasey) [inline src/js/utils/log.js]: do we want to log a warning if the default log level is invalid?

File: src/js/utils/log.js [hunk 12]
```diff
@@ -76,32 +83,127 @@ export const logByType = (type, args, stringify = !!IE_VERSION && IE_VERSION < 1
...
+  level = window.VIDEOJS_DEFAULT_LOG_LEVEL;
+}
+
+if (Object.defineProperty) {
+  Object.defineProperty(log.levels, 'DEFAULT', {value: level});
```
Comment (BrandonOCasey) [inline src/js/utils/log.js]: why not make `log.levels.DEFAULT = level;` the default and override it if `Object.defineProperty` exists?

File: docs/faq.md (hunk 9 above)
Comment (misteroneill) [inline docs/faq.md]: Absolutely. No idea how it got in here... weird.

File: src/js/utils/log.js (hunk 10 above)
Comment (misteroneill) [inline src/js/utils/log.js]: Eliminating the `defineProperties` usage altogether.

File: src/js/utils/log.js (hunk 11 above)
Comment (misteroneill) [inline src/js/utils/log.js]: Killing off this global entirely.

File: src/js/utils/log.js (hunk 12 above)
Comment (misteroneill) [inline src/js/utils/log.js]: Doing something simple like you're suggesting, but getting rid of the `defineProperty` crap altogether.

File: src/js/utils/log.js (hunk 6 above)
Comment (misteroneill) [inline src/js/utils/log.js]: No longer relevant as `window.VIDEOJS_DEFAULT_LOG_LEVEL` was removed.

File: docs/faq.md [hunk 13]
```diff
@@ -169,6 +169,13 @@ help but are outside of the scope of video.js.
 
 For content that must be highly secure [videojs-contrib-eme][eme] adds DRM support.
 
+## Q: Can I turn off video.js logging?
```
Comment (gkatsev) [inline docs/faq.md]: Should we list valid log levels here?

Comment (misteroneill) [inline docs/faq.md]: Added debugging guide, which includes that information. And added a link to that guide here.

File: docs/guides/debugging.md [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+By passing a string, the logging level can be changed to one of the available logging levels:
+
+```js
+videojs.log.level('error'); // show only error messages and suppress others
+videojs.log('foo'); // does nothing
+videojs.log.warn('foo'); // does nothing
+videojs.log.error('foo'); // logs "foo" as an error
+```
+
+Available logging levels are:
```
Comment (gkatsev) [inline docs/guides/debugging.md]: Can we make this a `####` heading? That way it shows up in the TOC as well.

Comment (misteroneill) [inline docs/guides/debugging.md]: Done.
